packages/loader/container-loader/src/protocol.ts

```diff
@@ -83,6 +89,14 @@ export class ProtocolHandler extends ProtocolOpHandler implements IProtocolHandl
             }
         }
 
+        const currentClientSeqNum = message.clientSequenceNumber;
+        if (client?.lastClientSequenceNumber !== undefined && !canBeCoalescedByService(message)) {
```

Comment:
So the problem here (based on [Teams thread](https://teams.microsoft.com/l/message/19:50292e8934024fc19d6ca2080dd7681e@thread.skype/1666651884725?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=9ce27575-2f82-4689-abdb-bcff07e8063b&parentMessageId=1666651884725&teamName=Fluid%20Framework&channelName=Dev&createdTime=1666651884725) is that we check current op, while we should check previous op that client received.
I.e. if previous op was noop, then it's Ok to have a gap in between that noop and this op

Reply:
no the problem was the client didn't receive that one noop either

Reply:
I think my suggestion is wrong, but first let me explain the process.
Noops can be coalesced by service. That means that if client sent 2 noops in a row, service may decide to drop one of them. It can do one of the following. It may delay sending noop, and eventually not send it if sending it does not carry any new information to the client. Noops are used to tell everyone about reference sequence number (last processed op) by a client who is sending such op. if MSN has already moved to that position or higher, then every connected client is at or above MSN, and thus its reference sequence number is at or above as well - sending noop will not convey any new info, so service skips sending it.

The suggestion I gave earlier is wrong, because it depends less on if this or previous op was noop. It depends only on - was the op that is missing a noop or not. And it's impossible to say that without tracking more state, i.e. tracking what was the last op sent. This layer simply does not have that info.

While everything is possible (it's software :)), I think it would be wrong to supply that info to this layer.
So I'd suggest abandoning this approach, and switching to more limited approach of client checking the same thing for only itself (not all clients), and thus moving this logic into DeltaManager.processInboundMessage() method.
There, we have access to all ops received (this method) and all ops sent by this client (submit()).

Given that noops can show up multiple times (independently) in outbound stream, it's hard to properly track that without tracking all pending (non acked ops). So I'd suggest simpler approach:
- sending ops: have a counter of number of noops we sent. Increment every time noop is sent.
- receiving ops:
   - If this counter is above zero, and we have a gap, decrease counter by size of gap
   - if we have a gap but counter is at zero - that's violation
- if we received all ops, counter should be zero (optional to check).

Reply:
just want to double check there is localOpsToIgnore in connectionManager, as well as other validation for clientSequenceNumber, would that be relevant?